### DISCUSSION TITLE
Integrate custom graphics of QGraphicsScene via QPainter

### DISCUSSION THREAD

**Discussion by ubruhin** [slint-ui/slint#5704] [General]
Hi,

I have a large C++/QtWidgets application ([LibrePCB](https://librepcb.org/)) making heavy use of [`QGraphicsScene`](https://doc.qt.io/qt-6/qgraphicsscene.html) and am evaluating alternative UI toolkits - in particular QtQuick and Slint. But for a step-by-step migration, I need a way to integrate the already existing `QGraphicsScene` since a rewrite of it would be too much effort.

In QtQuick I was able to build a PoC using [`QQuickPaintedItem`](https://doc.qt.io/qt-6/qquickpainteditem.html) which provides a `QPainter` where I can render the scene with [`QGraphicsScene::render()`](https://doc.qt.io/qt-6/qgraphicsscene.html#render). Works quite well so far.

If I understand correctly, the [Qt backend](https://releases.slint.dev/1.3.2/docs/slint/src/advanced/backend_qt) of Slint uses `QPainter` for drawing the whole window. So I wonder if it would be possible to draw on the window through this `QPainter`? Ideally in background, so that Slint controls can be used as overlay on the painted scene.

Is it possible to get access to the `QPainter` of the Qt backend from C++? Or any other idea how to integrate a `QGraphicsScene`?

Another story is whether Slint works together with a Qt application at all, as probably I need to run Qt's event loop somehow to get all the events etc. working - or would this even work out of the box with the Qt backend :thinking: EDIT: This seems to work indeed!

Thanks!

**Comment by tronical** [comment]
At this point we do not expose the `QPainter` the Qt backend uses. We do expose a `QWidget` in the C++ API, but that's not something I think we'd like to keep forever. Perhaps you could layer that with translucency on top of a `QGraphicsWidget`.

Generally speaking, composition of a scene rendered with Slint controls on top of another scene should always be possible one way or another.

**Reply by ubruhin** [reply]
Alright, thanks for the answer. I was able to put a `QGraphicsView` on the exposed `QWidget` (which would be the ideal solution, as `QGraphicsView` does exactly what we need). However, it is painted in foreground, so no Slint elements can be used as overlay :-( Also I'm not very happy with using an API which might be removed some day...

I was also able to render `QGraphicsScene` into a `slint::Image`, but here I'm not sure if that will lead to a performance bottleneck. At least I expect it to be much less efficient than directly drawing on the `QPainter` or so (several image buffer copies involved).

Since in https://github.com/slint-ui/slint/discussions/5714 there was just another request about a canvas, let me quickly show our use-case (`slint::Image` on the left, `QGraphicsView` on the right):

![librepcb-slint](https://github.com/user-attachments/assets/07488c0d-2264-4c1a-b09b-10edaae62c66)

For long term we want to migrate away from `QGraphicsScene` as we hit performance issues with large scenes - I think we'll need something GPU accelerated.

**Reply by tronical** [reply]
> Alright, thanks for the answer. I was able to put a `QGraphicsView` on the exposed `QWidget` (which would be the ideal solution, as `QGraphicsView` does exactly what we need). However, it is painted in foreground, so no Slint elements can be used as overlay :-(

I was thinking the other way around :). Use `background: transparent;` on the `Slint` `Window` element and place the `QWidget` on top of your `QQGraphicsView`.

**Reply by ubruhin** [reply]
Oh...hmm... you mean reparenting the `QWidget` :thinking: That sounds interesting, but I'm a bit concerned that the `QWidget` will no longer be the root window, which might break some functionality of the Qt backend :thinking: 

Anyway, I gave it a quick try:

```cpp
  QMainWindow win;
  win.resize(800, 600);

  QGraphicsView* view = new QGraphicsView(&win);
  view->setScene(...);

  QWidget* widget =
      static_cast<QWidget*>(slint::cbindgen_private::slint_qt_get_widget(
          &ui->window().window_handle()));
  widget->setParent(&win);
```

It works somehow, with some drawbacks:

- Both widgets need to be resized manually when the window is resized (ok, one of them can be done with Qt layouts).
- Input events are not forwarded to `QGraphicsView` as transparent background != transparent for input events. Could be worked around by placing a `TouchArea` over `QGraphicsView` and manually forwarding the events.
- There's some issue with repainting when switching between transparent and non-transparent content (resizing window removes the artifacts):
    ![librepcb-slint-repaint](https://github.com/user-attachments/assets/25f8abf2-aa02-4a74-9a4f-f2ea2c08d12b)

Overall this way feels very hacky to me, not really viable for a large, professional software. Will need to think about it if and how we could use Slint for the GUI. Thanks anyway for your help so far :+1:

**Reply by ogoffart** [reply]
Another option to look at is QGraphicsProxyWidget to put the Slint layer somehow within the graphicsscene.

Btw, to get the widget, you can use:

```c++
QWidget* widget = ui->qwidget();
```

https://releases.slint.dev/1.7.1/docs/cpp/api/classslint_1_1interpreter_1_1componentinstance#_CPPv4NK5slint11interpreter17ComponentInstance7qwidgetEv

Regarding the resizing, i think this is a bug, maybe https://github.com/slint-ui/slint/issues/5563 or https://github.com/slint-ui/slint/issues/3990 or some other.

**Reply by ubruhin** [reply]
I think `QGraphicsProxyWidget` can't help us since basically we need to embed a graphics scene (or in fact, *multiple* graphics scenes) into the Slint UI, not the other way around. Also the graphics scenes are not always there, they will be part of tabs while the outer UI (e.g. the panel on the left in my screenshots above) are static.

I wonder whether it would be possible to extend the `RenderingNotifier` callback to work also on non-GL backends, to allow native, backend-dependent painting in `BeforeRendering` or `AfterRendering` events :thinking:  With the Qt backend it could mean to pass the `QPainter*` to the application so we can draw custom underlays and overlays, the same way as we could do with OpenGL.

> Btw, to get the widget, you can use:
>
>    QWidget* widget = ui->qwidget();

Ah I didn't find that, thanks for the hint :+1:

**Reply by ogoffart** [reply]
Yes, i think it'd indeed make sense to extend the rendering notifier API to the Qt backend. 
(i don't think we need a shared QPainter. You can just pass a QWidget* and one can re-create a QPainter for it)

**Reply by ubruhin** [reply]
Oh, just for the record, I found another way which could be interesting:

```cpp
 class EventFilter : public QObject {
  public:
    EventFilter(slint::ComponentHandle<ui::AppWindow> app, GraphicsScene* scene)
      : QObject(), app(app), scene(scene) {}

    virtual bool eventFilter(QObject* watched, QEvent* event) override {
      if (event->type() == QEvent::Paint) {
        QPainter painter(static_cast<QWidget*>(watched));
        myGraphicsScene->render(&painter);
      }
      return QObject::eventFilter(watched, event);
    }
  };

auto app = ui::AppWindow::create();
QWidget* widget =
      static_cast<QWidget*>(slint::cbindgen_private::slint_qt_get_widget(
          &app->window().window_handle()));
widget->installEventFilter(new EventFilter());
```

This seems to work pretty well as far as I can see.

Btw, for some reason the `qwidget()` method doesn't exist in my `AppWindow` class (though I don't understand how it is related to the `ComponentInstance` class) :thinking: Using Slint 1.7.1 and the Qt backend is obviously there as the code above works fine.
